client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -768,11 +833,56 @@ private void setHeaders(HttpRequest req, Collection<Header> requestHeaders) {
                     req.addHeader(defaultHeader);
                 }
             }
+            // Add compatibility request headers if compatibility mode has been enabled
+            if (useCompatibility) {
+                addCompatibilityFor(req, entityHeader, "Content-Type");
+                addCompatibilityFor(req, entityHeader, "Accept");
+            }
             if (compressionEnabled) {
                 req.addHeader("Accept-Encoding", "gzip");
             }
         }
 
+        /**
+         * Go through all the request's existing headers, looking for {@code headerName} headers and if they exist,
+         * changing them to use version compatibility. If no request headers are changed, modify the entity type header if appropriate
+         */
+        private void addCompatibilityFor(HttpRequest req, Header entityHeader, String headerName) {
+            // Modify any existing "Content-Type" headers on the request to use the version compatibility, if available
+            boolean contentTypeModified = false;
+            for (Header header : req.getHeaders(headerName)) {
+                contentTypeModified = contentTypeModified || modifyHeader(req, header, headerName);
+            }
+
+            // If there were no request-specific headers, modify the request entity's header to be compatible
+            if (entityHeader != null && contentTypeModified == false) {
+                contentTypeModified = modifyHeader(req, entityHeader, headerName);
+            }
+
+            // If there were no changed headers at all, add the default compatibility header
+            if (contentTypeModified == false) {
```

Comment:
As I understand it, this will force `application/vnd.elasticsearch+json` if there was either no value for the header or if it wasn't one of the known mime types.

We should not change header values that aren't in the known ones: not sure if the case exists in ES, but a custom plugin may accept other content types.

Also we should not add a `Content-Type` header if there was none (requests without a body).
We should however force `Accept` to `application/vnd.elasticsearch+json;compatible-with=7` if it's missing as ES will fallback to json if there's no `Accept` header.

Additional note: ES also interprets `application/*` as `application/json`, so we may want to add this as another `EntityType` that translates to json.

Reply:
Ahh okay, I see now that it is okay to elide the `Content-Type` header for compatibility if there is no body, so I'll make that change.

I'll add the `*` entity type also, I didn't realize ES interpreted that as `application/json`, thanks for letting me know!